enclave_build/src/image_manager.rs

```diff
@@ -0,0 +1,178 @@
+                Ok(root_path) => {
+                    // Try to create/read the storage. If the storage could not be created, log the error
+                    OciStorage::new(&root_path)
+                        .map_err(|err| eprintln!("{:?}", err))
+                        .ok()
+                }
+                Err(_) => None,
+            };
+
+        let image_details = Self::fetch_image_details(&image_name, storage).await?;
+
+        Ok(Self {
+            image_name,
+            image_details,
+        })
+    }
+
+    /// Returns a struct containing image metadata.
+    ///
+    /// If the image is stored correctly, the function tries to fetch the image from the storage.
+    ///
+    /// If the image is not stored or a storage was not created (the 'storage' field is None),
+    /// it pulls the image, stores it (if the 'storage' field is not None) and returns its metadata.
+    ///
+    /// If the pull succeeded but the store operation failed, it returns the pulled image metadata.
+    async fn fetch_image_details(
+        image_name: &str,
+        mut storage: Option<OciStorage>,
+    ) -> Result<ImageDetails> {
+        let local_storage = storage.as_mut();
+
+        let image_details = if let Some(storage) = local_storage {
+            // Try to fetch the image from the storage
+            storage.fetch_image_details(image_name).map_err(|err| {
+                // Log the fetching error
+                eprintln!("{:?}", err);
+                err
+            })
+        } else {
+            Err(EnclaveBuildError::OciStorageNotFound(
+                "Local storage missing".to_string(),
+            ))
+        };
+
+        // If the fetching failed, pull it from remote and store it
+        match image_details {
+            Ok(details) => Ok(details),
+            Err(err) => {
+                // Log error
+                eprintln!("Fetching from storage failed: {}", err);
+
+                // The image is not stored, so try to pull and then store it
+                let image_data = crate::pull::pull_image_data(image_name).await?;
+
+                // If the store operation fails, discard error and proceed with getting the details
+                if let Some(local_storage) = storage.as_mut() {
+                    local_storage
+                        .store_image_data(image_name, &image_data)
+                        .map_err(|err| eprintln!("Failed to store image: {:?}", err))
+                        .ok();
+                }
+
+                // Get the image metadata from the pulled struct
+                let image_details = ImageDetails::build_details(image_name, &image_data)?;
+
+                Ok(image_details)
+            }
+        }
+    }
+}
+
+/// Adds the default ":latest" tag to an image if it is untagged
+fn normalize_tag(image_name: &str) -> String {
+    let name = image_name.to_string();
+    match name.contains(':') {
```

Comment:
Is any name that contains at least one ":" a valid tagged image? Is it valid to have "::" for example?

Reply:
Found a tag parser in `oci_distribution`